docs: update maintainer guidelines (#11376)
Homebrew/brew: docs/Homebrew-homebrew-core-Merge-Checklist.md

> # Core Maintainer Guide

Reviewer: ```suggestion
# Homebrew/homebrew-core Maintainer Guide
```

Author: I, personally, prefer the shorter names because the longer names all look the same at first glance, but I'll go with these if that's what's preferred.

## Quick merge checklist

A detailed checklist can be found [below](#detailed-merge-checklist). This is all that really matters:

> - Ensure the name seems reasonable.

Reviewer 2: ```suggestion
- Ensure the name seems reasonable. Formulae that are also packaged by other package managers (e.g. Debian, Ubuntu) should be named consistently (subject to minor differences due to Homebrew formula naming conventions).
```
Actually, maybe this needs to go into the `Naming` section below (with a link here to down there?).

  [gem](https://en.wikipedia.org/wiki/RubyGems),
  [cpan](https://en.wikipedia.org/wiki/Cpan) or
  [pip](https://pip.pypa.io/en/stable/).
- Ensure that any dependencies are accurate and minimal. We don't need to
  support every possible optional feature for the software.
- When bottles aren't required or affected, use the GitHub squash & merge workflow for a single-formula PR or rebase & merge workflow for a multiple-formulae PR. See [below](#how-to-merge-without-bottles) for more details.
- Use `brew pr-publish` or `brew pr-pull` otherwise, which adds messages to auto-close pull requests and pull bottles built by the Brew Test Bot.
- Thank people for contributing.

Checking dependencies is important, because they will probably stick around
forever. Nobody really checks if they are necessary or not.

Depend on as little stuff as possible. Disable X11 functionality if possible.
For example, we build Wireshark, but not the heavy GUI.

For [some formulae](https://github.com/Homebrew/homebrew-core/search?q=%22homebrew%2Fmirror%22&unscoped_q=%22homebrew%2Fmirror%22),
we mirror the tarballs to our own BinTray automatically as part of the

> bottle publish CI run.

Reviewer 2: We don't have `brew mirror` anymore.

is fine. If you aren’t an expert in the tool in question, you can’t really
gauge if the formula installed the program correctly. At some point an expert
will come along, cry blue murder that it doesn’t work, and fix it. This is how
open source works. Ideally, request a `test do` block to test that
functionality is consistently available.

If the formula uses a repository, then the `url` parameter should have a
tag or revision. `url`s have versions and are stable (not yet
implemented!).

Don't merge any formula updates with failing `brew test`s. If a `test do` block
is failing it needs to be fixed. This may involve replacing more involved tests
with those that are more reliable. This is fine: false positives are better than
false negatives as we don't want to teach maintainers to merge red PRs. If the
test failure is believed to be due to a bug in `Homebrew/brew` or the CI system,
that bug must be fixed, or worked around in the formula to yield a passing test,
before the PR can be merged.

## Duplicates

We now accept stuff that comes with macOS as long as it uses `keg_only :provided_by_macos` to be keg-only by default.

> ## Removing formulae

Reviewer 2: Now that we also `deprecate!` and `disable!` formulae, it would be good to outline some guidelines here for doing that too.

Author: I'll probably just link to https://docs.brew.sh/Deprecating-Disabling-and-Removing-Formulae

require new bottles to be pulled should use GitHub's squash & merge or rebase & merge workflows.
See the [table below](#how-to-merge-without-bottles) for more details.

Otherwise, you should use `brew pr-pull` (or `rebase`/`cherry-pick` contributions).

Don’t `rebase` until you finally `push`. Once `master` is pushed, you can’t
`rebase`: **you’re a maintainer now!**

Cherry-picking changes the date of the commit, which kind of sucks.

Don’t `merge` unclean branches. So if someone is still learning `git` and
their branch is filled with nonsensical merges, then `rebase` and squash
the commits. Our main branch history should be useful to other people,
not confusing.

Here’s a flowchart for managing a PR which is ready to merge:

![Flowchart for managing pull requests](assets/img/docs/managing-pull-requests.drawio.svg)

Only one maintainer is necessary to approve and merge the addition of a new or updated formula which passes CI. However, if the formula addition or update proves controversial the maintainer who adds it will be expected to answer requests and fix problems that arise with it in future.

### How to merge without bottles

> Here are guidelines about when to use squash & merge versus rebase & merge. These options should only be used with PRs where bottles are not needed or affected.

Reviewer: ```suggestion
Here are guidelines about when to use squash & merge versus rebase & merge. These options should only be used with PRs where bottles are not affected.
```

Because 🔜 bottles will always be needed.

> | | PR modified a single formula | PR modifies multiple formulae |

Reviewer 3: ```suggestion
| | PR modifies a single formula | PR modifies multiple formulae |
```

is failing it needs to be fixed. This may involve replacing more involved tests
with those that are more reliable. This is fine: false positives are better than
false negatives as we don't want to teach maintainers to merge red PRs. If the
test failure is believed to be due to a bug in `Homebrew/brew` or the CI system,
that bug must be fixed, or worked around in the formula to yield a passing test,
before the PR can be merged.

## Duplicates

We now accept stuff that comes with macOS as long as it uses `keg_only :provided_by_macos` to be keg-only by default.

## Removing formulae

Formulae that:

- work on at least 2/3 of our supported macOS versions in the default Homebrew prefix
- do not require patches rejected by upstream to work
- do not have known security vulnerabilities or CVEs for the version we package
- are shown to be still installed by users in our analytics with a `BuildError` rate of <25%

should not be removed from Homebrew. The exception to this rule are [versioned formulae](Versions.md) for which there are higher standards of usage and a maximum number of versions for a given formula.

> For more information about deprecating, disabling and removing formulae, see the
> [Deprecating, Disabling, and Removing Formulae page](Deprecating-Disabling-and-Removing-Formulae.md)

Reviewer 3: ```suggestion
For more information about deprecating, disabling and removing formulae, see
[Deprecating, Disabling, and Removing Formulae](Deprecating-Disabling-and-Removing-Formulae.md).
```

## Detailed merge checklist

The following checklist is intended to help maintainers decide on
whether to merge, request changes or close a PR. It also brings more

> transparency for contributors in addition to

Reviewer 3: ```suggestion
transparency for contributors in addition to the
```

# Homebrew/homebrew-core Maintainer Guide

## Quick merge checklist

A detailed checklist can be found [below](#detailed-merge-checklist). This is all that really matters:

- Ensure the name seems reasonable.
- Add aliases.
- Ensure it uses `keg_only :provided_by_macos` if it already comes with macOS.
- Ensure it is not a library that can be installed with
  [gem](https://en.wikipedia.org/wiki/RubyGems),
  [cpan](https://en.wikipedia.org/wiki/Cpan) or

>   [pip](https://pip.pypa.io/en/stable/).

Reviewer 4: Do we actually check for this?

Reviewer 5: HB ships at least a few `pip`-installable tools. The most well-known is probably `youtube-dl`.

But even the ones I know of which are libraries provide command-line interfaces, so it depends on the interpretation of this section.

I’d be for continuing to provide those. HB’s usefulness grows if a user doesn’t have to learn/use/update a language’s package manager to keep a single popular tool up-to-date.

Reviewer: Yeh, hence the "library" bit

Reviewer 4: We've recently started shipping `six`, which is a pip installable library to make packaging easier. I don't think this is still actually enforced (and I don't see any good reasons why we would)

Reviewer: If we need it for our own benefit: cool/fine. The intent is that we don't have a million Python libraries and duplicate that package manager (like e.g. Debian does)

Reviewer 4: I thought `apt` was your dream package manager 😄

- Ensure that any dependencies are accurate and minimal. We don't need to
  support every possible optional feature for the software.
- When bottles aren't required or affected, use the GitHub squash & merge workflow for a single-formula PR or rebase & merge workflow for a multiple-formulae PR. See the ["How to merge without bottles" section below](#how-to-merge-without-bottles)
for more details.
- Use `brew pr-publish` or `brew pr-pull` otherwise, which adds messages to auto-close pull requests and pull bottles built by the Brew Test Bot.
- Thank people for contributing.

Checking dependencies is important, because they will probably stick around
forever. Nobody really checks if they are necessary or not.

Depend on as little stuff as possible. Disable X11 functionality if possible.
For example, we build Wireshark, but not the heavy GUI.

Homebrew is about Unix software. Stuff that builds to an `.app` should
be in Homebrew Cask instead.

## Merging, rebasing, cherry-picking

> In most cases, you can simply approve a PR and an automatic merge (with bottles)

Reviewer 4: ```suggestion
In most cases modifying a formula, you can simply approve a PR and an automatic merge (with bottles)
```

Reviewer 3: "In most cases of modifying a formula"

Reviewer 4: I pick the second one
```suggestion
In most cases that modify a formula, you can simply approve a PR and an automatic merge (with bottles)
```

Author: How about:
```suggestion
For most PRs that make formula modifications, you can simply approve the PR and an automatic merge (with bottles)
```

will be performed by [@BrewTestBot](https://github.com/BrewTestBot).
See [Brew Test Bot For Core Contributors](Brew-Test-Bot-For-Core-Contributors.md) for more information.

PRs modifying formulae that don't need bottles or making changes that don't
require new bottles to be pulled should use GitHub's squash & merge or rebase & merge workflows.
See the [table below](#how-to-merge-without-bottles) for more details.

> Otherwise, you should use `brew pr-pull` (or `rebase`/`cherry-pick` contributions).

Reviewer 4: ```suggestion
Otherwise, for example for new formulae, you should use `brew pr-pull` (or `rebase`/`cherry-pick` contributions).
```

Author: Is this okay?

```suggestion
Otherwise (e.g. for PRs adding new formula), you should use `brew pr-pull` (or `rebase`/`cherry-pick` contributions).
```

Author: Actually, this should probably mention `brew pr-publish` before `brew pr-pull`, right?
